src/track/serato/beatgrid.cpp

```diff
@@ -417,6 +417,97 @@ QByteArray SeratoBeatGrid::dumpBase64Encoded() const {
+        previousBeatPositionSamples = currentBeatPositionSamples;
+        currentBeatPositionSamples = static_cast<SINT>(pBeatsIterator->next());
+
+        const double currentDeltaSamples = currentBeatPositionSamples - previousBeatPositionSamples;
+        if (currentDeltaSamples != previousDeltaSamples) {
+            if (!nonTerminalMarkers.isEmpty()) {
+                DEBUG_ASSERT(beatsTillNextMarker > 0);
+                const auto pNonTerminalMarker =
+                        nonTerminalMarkers.at(nonTerminalMarkers.size() - 1);
+                DEBUG_ASSERT(pNonTerminalMarker);
+                pNonTerminalMarker->setBeatsTillNextMarker(beatsTillNextMarker);
+
+                // After adding the first marker, the the sample delta won't
+                // match, because it compares the distance between beats 1 and
+                // two with the distance between the start of the track and the
+                // first beat. This special case makes sure that we don't add
+                // an unnecessary additional marker that has
+                // beatsTillNextMarker set to 1.
+                if (nonTerminalMarkers.size() == 1) {
+                    previousDeltaSamples = currentDeltaSamples;
+                    beatsTillNextMarker++;
+                    continue;
+                }
+            }
+            beatsTillNextMarker = 0;
+
+            // Don't create a SeratoBeatGridNonTerminalMarker entry for the
+            // last beat, this needs to be a terminal marker entry
+            if (pBeatsIterator->hasNext()) {
+                const double positionSecs =
+                        streamInfo.getSignalInfo().frames2secs(
+                                streamInfo.getSignalInfo().samples2frames(
+                                        currentBeatPositionSamples)) -
+                        timingOffsetSecs;
+                nonTerminalMarkers.append(
+                        std::make_shared<SeratoBeatGridNonTerminalMarker>(
+                                positionSecs, 0));
+            }
+        }
+        previousDeltaSamples = currentDeltaSamples;
+        beatsTillNextMarker++;
+    }
+
+    VERIFY_OR_DEBUG_ASSERT(currentBeatPositionSamples != -1) {
+        return;
+    }
+
+    DEBUG_ASSERT(!nonTerminalMarkers.isEmpty());
+    DEBUG_ASSERT(beatsTillNextMarker > 0);
+    const auto pNonTerminalMarker = nonTerminalMarkers.at(nonTerminalMarkers.size() - 1);
+    DEBUG_ASSERT(pNonTerminalMarker);
+    pNonTerminalMarker->setBeatsTillNextMarker(beatsTillNextMarker);
+
+    const double positionSecs =
+            streamInfo.getSignalInfo().frames2secs(
+                    streamInfo.getSignalInfo().samples2frames(
+                            currentBeatPositionSamples)) -
+            timingOffsetSecs;
+    const double bpm = pBeats->getBpmAroundPosition(currentBeatPositionSamples, 1);
+    setTerminalMarker(std::make_shared<SeratoBeatGridTerminalMarker>(positionSecs, bpm));
```

Comment:
Unrelated: Why are both beat grid markers shared pointers? They could be simple, immutable value classes. Their content is not bigger than a pointer. I don't see a reason why it would be necessary to allocate them on the heap and share references to them instead of simple copying their values?

Reply:
> Their content is not bigger than a pointer.

That is only true in 64 bit systems. What about 32-bit ARM? Or Windows?

Reply:
It is still only a few bytes. Keeping them in a contiguous piece of memory is more cache-friendly. If you do not actually need to share those pointers I would suggest to simply store the values in place.

Reply:
I'll take care of that in another PR after this has been merged.